stdlib/strings/strings.go

```diff
@@ -382,7 +385,77 @@ func generateUnicodeIsFunction(name string, Fn func(rune) bool) values.Function
 	)
 }
 
+var strlen = values.NewFunction(
+	"strlen",
+	semantic.NewFunctionPolyType(semantic.FunctionPolySignature{
+		Parameters: map[string]semantic.PolyType{stringArgV: semantic.String},
+		Required:   semantic.LabelSet{stringArgV},
+		Return:     semantic.Int,
+	}),
+	func(args values.Object) (values.Value, error) {
+		v, ok := args.Get(stringArgV)
+		if !ok {
+			return nil, fmt.Errorf("missing argument %q", stringArgV)
+		}
+
+		if v.Type().Nature() == semantic.String {
+			return values.NewInt(int64(utf8.RuneCountInString(v.Str()))), nil
+		}
+
+		return nil, fmt.Errorf("procedure cannot be executed")
+	}, false,
+)
+
+var substring = values.NewFunction(
+	"substring",
+	semantic.NewFunctionPolyType(semantic.FunctionPolySignature{
+		Parameters: map[string]semantic.PolyType{
+			stringArgV: semantic.String,
+			start:      semantic.Int,
+			end:        semantic.Int},
+		Required: semantic.LabelSet{stringArgV, start, end},
+		Return:   semantic.String,
+	}),
+	func(args values.Object) (values.Value, error) {
+		v, vOk := args.Get(stringArgV)
+		a, aOk := args.Get(start)
+		b, bOk := args.Get(end)
+		if !aOk || !bOk || !vOk {
+			return nil, fmt.Errorf("missing argument")
+		}
+
+		vStr := v.Str()
+		aInt := int(a.Int())
+		bInt := int(b.Int())
+		if aInt < 0 || bInt > len(vStr) {
+			return nil, fmt.Errorf("indices out of bounds")
+		}
+
+		if (v.Type().Nature() == semantic.String) && (a.Type().Nature() == semantic.Int) && (b.Type().Nature() == semantic.Int) {
+			newStr := ""
+			count := 0
+			for _, c := range vStr {
+				if count < aInt {
+					count++
+					continue
+				}
+				if count >= bInt {
+					break
+				}
+				newStr = strings.Join([]string{newStr, string(c)}, "")
```

Comment:
This line is still constructing a new string with each loop which is allocation heavy. You can use Go's substring here to help.

I am thinking something along the lines of finding the byteStart and byteEnd indexes and then doing this after the loop.

```
return values.NewString(vStr[byteStart:byteEnd]), nil
```
Can you try that and use [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp) to compare the two implementations to see which is better?

Reply:
using `byteStart` and `byteEnd` makes it run about the same or faster, depending on which end of the string I'm getting the substring from. will be changing it to the faster version